rfcs/0001-buildpack-order.md

```diff
@@ -0,0 +1,60 @@
+# Proposed Buildpack Ordering in a Builder
+
+## Summary
+
+`pack inspect-builder gcr.io/paketo-buildpack/builder` should display a `Detection Order` that is deterministic and will never make it impossible for any buildpack to detect.
+
+## Motivation
+
+After a change to the builder detection order it became impossible for `php` apps to detect, because `nginx` and `httpd` were before `php` in the detect order. 
+The `httpd.conf` or `nginx.conf` included in `php` apps will cause the `nginx` or `http` buildpacks to detect `true` before `php` ever gets to run `detect`,
+making it impossible to access the `php` buildpack.
+
+
+## Detailed Explanation
+
+We propose the following rule to order buildpacks in the builder's `Detect Order`:
+
+- All meta buildpacks should be come before implementation buildpacks, and sort each group by buildpaco-id.
+
+- To prove this will always be valid: (1)
+    - Consider 2 buildpacks A and B.
+    - If any of buildpack B's order groups are a subset of any of buildpack A's order groups, then buildpack A must come before buildpack B in the builder order. 
+    - An implementation buildpack can be defined as an order group of itself.
+
+Thus, our exiting buildpacks would be ordered as follows:
```

Comment:
Would be great to understand our motivation from moving away from the CF ordering: https://github.com/cloudfoundry/cf-deployment/blob/41a60509ffd73662e6fac059ff39ef86b913e48e/cf-deployment.yml#L874-L896

Many of those are ordered carefully to support common app configurations. E.g., node is below Ruby and .NET Core, because Ruby and .NET Core apps often have package.json.

Reply:
Right, we were trying to come up with a more universal rule that could be validated, without a ton of knowledge of the individual languages, and I didn't know enough about the old ordering to compare its benefits.

That being said those are cases we didn't think about and it might be too hard or impossible to make a language agnostic rue in the way we did.

Reply:
I'm skeptical that this can be solved with a mechanical lexical ordering.  I _believe_ that we're currently in a situation where the detection criteria are completely disjoint and precedence doesn't matter, but I don't believe that that is a stable assumption to make for the long term.